tools/pubsys/src/repo/validate_repo/mod.rs

```diff
@@ -0,0 +1,252 @@
+            path: root_role_path,
+        })?,
+        datastore: workdir.path(),
+        metadata_base_url: metadata_url.as_str(),
+        targets_base_url: targets_url.as_str(),
+        limits: Limits::default(),
+        expiration_enforcement: ExpirationEnforcement::Safe,
+    };
+
+    // Load the repository
+    let repo = Repository::load(transport, settings).context(repo_error::RepoLoad {
+        metadata_base_url: metadata_url.clone(),
+    })?;
+    info!("Loaded TUF repo: {}", metadata_url);
+    // Try retrieving listed targets
+    info!(
+        "Attempting to download {}% of listed targets from {} for validation purposes",
+        percent_target_files, metadata_url
+    );
+    retrieve_percentage_of_targets(&repo, percent_target_files).await?;
+
+    Ok(())
+}
+
+/// Common entrypoint from main()
+pub(crate) async fn run(args: &Args, validate_repo_args: &ValidateRepoArgs) -> Result<(), Error> {
+    info!(
+        "Using infra config from path: {}",
+        args.infra_config_path.display()
+    );
+    let infra_config =
+        InfraConfig::from_path(&args.infra_config_path).context(repo_error::Config)?;
+    trace!("Parsed infra config: {:?}", infra_config);
+    let repo_config = infra_config
+        .repo
+        .as_ref()
+        .context(repo_error::MissingConfig {
+            missing: "repo section",
+        })?
+        .get(&validate_repo_args.repo)
+        .context(repo_error::MissingConfig {
+            missing: format!("definition for repo {}", &validate_repo_args.repo),
+        })?;
+    ensure!(
+        validate_repo_args.percent_target_files <= 100,
+        error::InvalidPercentage {
+            percentage: validate_repo_args.percent_target_files
+        }
+    );
+
+    let transport = RepoTransport::default();
+    let repo_urls = repo_urls(
+        &repo_config,
+        &validate_repo_args.variant,
+        &validate_repo_args.arch,
+    )?
+    .with_context(|| repo_error::MissingRepoUrls {
+        repo: &validate_repo_args.repo,
+    })?;
+    validate_repo(
+        &transport,
+        &validate_repo_args.root_role_path,
+        repo_urls.0,
+        repo_urls.1,
+        validate_repo_args.percent_target_files,
+    )
+    .await
+}
+
+mod error {
+    use snafu::Snafu;
+    use std::io;
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility = "pub(super)")]
+    pub(crate) enum Error {
+        #[snafu(display("Invalid percentage specified: {} is greater than 100", percentage))]
+        InvalidPercentage { percentage: u8 },
+
+        #[snafu(context(false), display("{}", source))]
+        Repo { source: crate::repo::Error },
+
+        #[snafu(display("Failed to download and write target '{}': {}", target, source))]
+        TargetDownload { target: String, source: io::Error },
+
+        #[snafu(display("Missing target: {}", target))]
+        TargetMissing { target: String },
+
+        #[snafu(display("Failed to spawn threads for fetching target: {}", source))]
```

Comment:
```suggestion
        #[snafu(display("Failed to spawn task for fetching target: {}", source))]
```
(tokio decides whether a task will be in a new thread or the current one)